[![Build Status](https://travis-ci.com/dariober/cnv_facets.svg?branch=master)](https://travis-ci.com/dariober/cnv_facets)
[![Language](https://img.shields.io/badge/language-R-brightgreen.svg)](https://cran.r-project.org/)
[![License](http://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/dariober/cnv_facets)

Somatic copy variant caller for next generation sequencing data based on the
[FACETS](https://github.com/mskcc/facets) package


<!-- vim-markdown-toc GFM -->

* [Requirements and Installation](#requirements-and-installation)
* [Input](#input)
    * [Option 1: BAM & VCF input](#option-1-bam--vcf-input)
    * [Option 2: Pileup input](#option-2-pileup-input)
* [Output](#output)
* [References](#references)

<!-- vim-markdown-toc -->

**cnv_facets.R** detects somatic Copy Number Variants (CNVs) from next
generation sequencing data such as *whole genome*, *whole exome* and *targeted
sequencing* experiments. In addition, it estimates tumour purity and ploidy.
cnv_facets.R is based on the original package [FACETS](https://github.com/mskcc/facets)

The advantage of **cnv_facets.R** over the original [FACETS](https://github.com/mskcc/facets) 
package is in the convenience of executing all the necessary steps, from BAM input to VCF 
output, in a single command line call.

Requirements and Installation
=============================

cnv_facets.R requires a reasonably recent version of
[R](https://cran.r-project.org/) on a `*Nix` operating system. At the time of
this writing, it has been developed and deployed on R 3.5 on CentOS 7.

To compile and install execute:

```
bash setup.sh --bin_dir </dir/on/path>
```

Where `/dir/on/path` is a directory on your PATH where you have permission to
write, *e.g.*, `~/bin`.

`setup.sh` accomplishes three main tasks:

* Install any missing, required R package

* Compile the helper program `snp-pileup` and move it to the designated PATH
  directory unless already found on PATH

* Run the test suite


Input
=====

Option 1: BAM & VCF input
-------------------------

* A bam file of the **tumour** sample

* A bam file of the **normal** sample (matched to the tumour, typically a blood
  sample from the same patient)

* A VCF file of common, polymorphic SNPs. For human samples, a good source is
  the dbSNP file
  [common_all.vcf.gz](https://www.ncbi.nlm.nih.gov/variation/docs/human_variation_vcf/). 
  See also NCBI [human variation sets in VCF Format](https://www.ncbi.nlm.nih.gov/variation/docs/human_variation_vcf/).

**USAGE**

```
cnv_facets.R -t <tumour.bam> -n <normal.bam> -vcf <snps.vcf.gz> -o <output_prefix> [...]
```

Option 2: Pileup input
----------------------

* A pileup file, comma separated, of read counts for the reference and alternate allele
  at polymorphic SNPs. This file must have the following columns (order of
  columns is not important, additional columns are ignored):
 
 * *Chromosome* Chromosome of the SNP
 
 * *Position* Position of the SNP
 
 * *File1R* Read depth supporting the REF allele in **normal** sample
 
 * *File1A* Read depth supporting the ALT allele in **normal** sample
 
 * *File2R* Read depth supporting the REF allele in **tumour** sample
 
 * *File2A* Read depth supporting the ALT allele in **tumour** sample

These are the first lines of the test file `test/data/stomach.csv.gz`
accompanying the original facets package:

```
"Chromosome","Position","Ref","Alt","File1R","File1A","File1E","File1D","File2R","File2A","File2E","File2D"
1,69424,N,N,170,117,0,0,158,103,0,0
1,69515,N,N,0,76,0,0,0,77,0,0
1,69536,N,N,103,0,0,0,99,0,0,0
1,808866,N,N,96,0,0,0,133,0,0,0
1,809120,N,N,66,0,0,0,105,0,0,0
```

This pileup file is generated by `cnv_facets.R` when run with bam input as in
option 1. Alternatively, it can be created by running `snp-pileup`. Using the
pileup file as input instead of the bam files has the advantage of saving
computing time.

**USAGE**

```
cnv_facets.R -p <pileup.csv.gz> -o <output_prefix> [...]
```

Output
======

References
==========
